### ISSUE TITLE
DGCRN的结果复现有出入

### ISSUE THREAD

**Issue by ZhangYuan1998** [zezhishao/BasicTS#59]
您好！Zezhi Shao,很感谢您开源的代码，这对我很有帮助！
但是，我在PEMS数据集上复现DGCRN时，发现按照代码的基础参数，与您图表里的结果相差甚远。
想在此咨询下您这个问题，是不是DGCRN一些设置没在开源代码中更新。
这个是PEMS08的最终结果练日志：
![image](https://github.com/zezhishao/BasicTS/assets/102887477/9dfb149e-66dd-476b-ba38-94792b1aead7)
PEMS04:
![image](https://github.com/zezhishao/BasicTS/assets/102887477/f476213e-1e3b-4a02-9e31-d6f05eeea377)
PEMS03:
![image](https://github.com/zezhishao/BasicTS/assets/102887477/54e73afe-28a2-4a13-aee5-d9c157c3057a)
期待您的回复！祝您万事顺遂！

**Comment by zezhishao** [comment]
另外，为了充分地训练模型，BasicTS默认是不支持Early Stopping的。所以您需要搜索最后一个`best_`，来获取最好的结果，而log日志中最后一轮的结果不是最好的结果。

**Comment by ZhangYuan1998** [comment]
非常感谢您及时的回复：
按照您的方式索引下，还是有些差距，应该是某些地方出现了问题，但是我确实是没有改动您目前开源的代码和配置，PEMS04:
![image](https://github.com/zezhishao/BasicTS/assets/102887477/638cf540-c839-46e8-ba68-1881fa81843b)
PEMS08:
![image](https://github.com/zezhishao/BasicTS/assets/102887477/9737e588-f108-460e-a0e8-8cab55d9fb5e)
PEMS08的配置如下：
2023-07-18 11:47:27,615 - easytorch-launcher - INFO - Launching EasyTorch training.
2023-07-18 11:47:28,007 - easytorch-env - INFO - Use GPUs 0.
2023-07-18 11:47:28,059 - easytorch-launcher - INFO - Initializing runner "<class 'basicts.runners.runner_zoo.dgcrn_runner.DGCRNRunner'>"
2023-07-18 11:47:28,059 - easytorch-env - INFO - Disable TF32 mode
2023-07-18 11:47:28,061 - easytorch - INFO - Set ckpt save dir: 'checkpoints/DGCRN_200/49307fcd8b1552e914ec9e5d31966f4b'
2023-07-18 11:47:28,061 - easytorch - INFO - Building model.
2023-07-18 11:47:32,536 - easytorch-training - INFO - Initializing training.
2023-07-18 11:47:32,536 - easytorch-training - INFO - Set clip grad, param: {'max_norm': 5.0}
2023-07-18 11:47:32,536 - easytorch-training - INFO - Building training data loader.
2023-07-18 11:47:32,615 - easytorch-training - INFO - Set optim: Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.001
    maximize: False
    weight_decay: 0.0001
)
2023-07-18 11:47:32,616 - easytorch-training - INFO - Set lr_scheduler: <torch.optim.lr_scheduler.MultiStepLR object at 0x7f9353031ee0>
2023-07-18 11:47:34,707 - easytorch-training - INFO - Initializing validation.
2023-07-18 11:47:34,707 - easytorch-training - INFO - Building val data loader.
2023-07-18 11:47:34,870 - easytorch-training - INFO - Epoch 1 / 200

  0%|          | 0/335 [00:00<?, ?it/s]DESCRIPTION: DGCRN model configuration
RUNNER: <class 'basicts.runners.runner_zoo.dgcrn_runner.DGCRNRunner'>
DATASET_CLS: <class 'basicts.data.dataset.TimeSeriesForecastingDataset'>
DATASET_NAME: PEMS08
DATASET_TYPE: Traffic flow
DATASET_INPUT_LEN: 12
DATASET_OUTPUT_LEN: 12
GPU_NUM: 1
NULL_VAL: 0.0
ENV:
  SEED: 1
  CUDNN:
    ENABLED: True
MODEL:
  NAME: DGCRN
  ARCH: <class 'basicts.archs.arch_zoo.dgcrn_arch.dgcrn_arch.DGCRN'>
  PARAM:
    gcn_depth: 2
    num_nodes: 170
    predefined_A: [tensor([[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.5000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 1.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        ...,
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.2500, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.1667, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000]]), tensor([[0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.5000,  ..., 0.0000, 0.0000, 0.0000],
        [0.0000, 1.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000],
        ...,
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.2500, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.1667, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000,  ..., 0.0000, 0.0000, 0.0000]])]
    dropout: 0.3
    subgraph_size: 20
    node_dim: 40
    middle_dim: 2
    seq_length: 12
    in_dim: 2
    list_weight: [0.05, 0.95, 0.95]
    tanhalpha: 3
    cl_decay_steps: 4000
    rnn_size: 64
    hyperGNN_dim: 16
  FORWARD_FEATURES: [0, 1]
  TARGET_FEATURES: [0]
TRAIN:
  LOSS: masked_mae
  OPTIM:
    TYPE: Adam
    PARAM:
      lr: 0.001
      weight_decay: 0.0001
  LR_SCHEDULER:
    TYPE: MultiStepLR
    PARAM:
      milestones: [100, 150]
      gamma: 0.5
  CLIP_GRAD_PARAM:
    max_norm: 5.0
  NUM_EPOCHS: 200
  CKPT_SAVE_DIR: checkpoints/DGCRN_200
  DATA:
    DIR: datasets/PEMS08
    BATCH_SIZE: 32
    PREFETCH: False
    SHUFFLE: True
    NUM_WORKERS: 2
    PIN_MEMORY: False
  CL:
    WARM_EPOCHS: 0
    CL_EPOCHS: 6
    PREDICTION_LENGTH: 12
VAL:
  INTERVAL: 1
  DATA:
    DIR: datasets/PEMS08
    BATCH_SIZE: 64
    PREFETCH: False
    SHUFFLE: False
    NUM_WORKERS: 2
    PIN_MEMORY: False
TEST:
  INTERVAL: 1
  DATA:
    DIR: datasets/PEMS08
    BATCH_SIZE: 64
    PREFETCH: False
    SHUFFLE: False
    NUM_WORKERS: 2
    PIN_MEMORY: False

train len: 10700
val len: 3567
test len: 3566

**Comment by ZhangYuan1998** [comment]
我后续也会继续实验下，看看这个问题出在什么地方，非常感谢您及时的回复！

**Comment by zezhishao** [comment]
好的，我的结果跑出来之后会回复您。

**Comment by zezhishao** [comment]
另外，您的BasicTS是最新版的吗？

**Comment by zezhishao** [comment]
非常感谢您的Issue，帮助我找到了`dgcrn_runner`中的一个bug。我已经在最新的commit中修复了该问题，您现在可以正确地复现dgcrn的性能了。这个bug应该只影响dgcrn。

**Comment by ZhangYuan1998** [comment]
好的！再次感谢您开源的代码以及非常及时的回复！
